app/models/highlight.rb

```diff
@@ -16,54 +21,113 @@ class Highlight < ApplicationRecord
                         :highlighted_content,
                         :location_strategies
 
-  validate :source_ids_uuid_for_openstax_page
+  validate :must_have_source_type_specific_ids
+  validate :must_have_consistent_neighbors, on: :create
 
   enum source_type: [:openstax_page]
 
-  before_validation :normalize_color, :normalize_source_ids
-
   scope :by_source_type, ->(source_type) { where(source_type: source_type) }
   scope :by_color, ->(color) { where(color: color) }
   scope :by_user, ->(user_id) { where(user_uuid: user_id) }
-  scope :by_order, ->(order_direction) {
-    order_direction = order_direction.to_sym
-    order(order_in_source: order_direction, source_order: order_direction)
-  }
-  scope :by_source_parent_ids, ->(source_parent_ids) do
-    where('source_parent_ids && ?', postgres_style_array(source_parent_ids))
-  end
+  scope :by_scope_id, ->(scope_id) { where(scope_id: scope_id) }
+  scope :by_source_ids, ->(source_ids) { where(source_id: source_ids) }
+
+  around_create :insert_new_highlight_between_neighbors
+  before_destroy :reconnect_neighbors_around_destroyed_highlight
 
   def normalize_color
     self.color = color&.downcase
   end
 
-  def normalize_source_ids
+  def normalize_ids
     if openstax_page?
       self.source_id = source_id&.downcase&.strip
-      source_parent_ids.map!(&:downcase).map!(&:strip)
+      self.scope_id = scope_id&.downcase&.strip
     end
   end
 
-  private
+  protected
+
+  def all_from_scope_and_source
+    Highlight.where(scope_id: scope_id).where(source_id: source_id)
+  end
 
   def valid_uuid?(uuid)
     VALID_UUID.match?(uuid.to_s.downcase)
   end
 
-  def source_ids_uuid_for_openstax_page
-    return unless openstax_page?
+  def validate_uuid(field, is_source_type_specific:)
+    unless VALID_UUID.match?(send(field).to_s.downcase)
+      errors.add(field, "must be a UUID#{' for source_type ' + source_type if is_source_type_specific}")
+    end
+  end
+
+  def must_have_source_type_specific_ids
+    if openstax_page?
+      validate_uuid(:scope_id, is_source_type_specific: true)
```

Comment:
this seems a bit confusing.  would we ever validate a UUID when not source_type specific?  this is already inside the if openstax_page?, so it's just validate_uuid right? later on... we'd have a `validate_id` or something right?

Reply:
Makes sense.  Changed.